### PR TITLE
[FIX] l10n_*: fix tests + don't decorate test class with freeze_time

### PR DESCRIPTION
**Pull request by JulienVR**
*: l10n_fr_fec, l10n_account_edi_ubl_cii_tests

Fix two failing tests in: - l10n_fr_fec: Add missing factor_percent in
the creation of a tax in the test for FEC - account_edi_ubl_cii: The
invoice name supposed the old convention per month while this was
changed to on a per year basis anyways for customer invoices.

These two tests were ignored by runbot for a while and began to fail
only after the runbot exception was removed (see:
https://runbot.odoo.com/runbot/build/22172258).

In addition, tests in `TestItEdi` started to fail as well when the
exception was removed ! In particular, if an exception was raised in the
`setUpClass` of `TestAccountFrFec`, the tests in `TestItEdi` were
failing.

This is because `TestAccountFrFec` is decorated with
`freeze_time('2021-05-02')`. Indeed, when one decorates a subclass of
`unittest.TestCase` with freeze_time, if the `setUpClass` raises an
Exception, `_freeze_time.stop()` is not called, and the frozen date is
propagated in the other test classes. As a consequence, when an
exception was raised in the `setUpClass` of `TestAccountFrFec`, the
freeze_time set in `TestAccountFrFec` (2021-05-02) was kept in
`TestItEdi`, thus causing tests to fail in `TestItEdi`.

As a conclusion, `freeze_time` should never decorate an Odoo test class.

### CODE REVIEW

**Comment by JulienVR** [comment]
linked to https://github.com/odoo/odoo/pull/108678
ping @jco-odoo

**Review by Xavier-Do — APPROVED** [review]
Thanks for the work and the summary of the issue!

I wont block the pr for that, just small details and tips for the future about [commit guidelines](https://www.odoo.com/documentation/16.0/contributing/development/coding_guidelines.html#git):

> [TAG] module: describe your change in a short sentence (ideally < 50 chars)

There are also good advices in [github commit guidelines](https://gist.github.com/tonibardina/9290fbc7d605b4f86919426e614fe692): 

I know that it is difficult to combine less than 50 and listing modules, but at least trying to have less than 72 is a good idea.
Having a too long title will lead to a line break making it quite difficult to read
image

Note: we accept having the modules listed on the next (third) line if there is too much, using a star in the tittle. There is no well defined guideline for that AFAIK but here is an [exemple](https://github.com/odoo/enterprise/commit/7704fe0ec99950cd7404fbd7b32965bfc64ed0c4):

```
[FIX] hr_*: fix post-always edit
    
    *: hr_appraisal, hr_referral, l10n_be_hr_payroll
```
Or like [this](https://github.com/odoo/enterprise/commit/7785843ed0b4152b082b1254fe9c82b1fa40fa20) with a fnmatch logic

Also, if you strictly respect the line break after 72, github will detect that as a fake line break and display it in on line when generating the pr message with is convenient when you are used to it ;).

**Comment by JulienVR** [comment]
> Thanks for the work and the summary of the issue!
> 
> I wont block the pr for that, just small details and tips for the future about [commit guidelines](https://www.odoo.com/documentation/16.0/contributing/development/coding_guidelines.html#git):
> 
> > [TAG] module: describe your change in a short sentence (ideally < 50 chars)
> 
> There are also good advices in [github commit guidelines](https://gist.github.com/tonibardina/9290fbc7d605b4f86919426e614fe692):
> 
> I know that it is difficult to combine less than 50 and listing modules, but at least trying to have less than 72 is a good idea. Having a too long title will lead to a line break making it quite difficult to read image
> 
> Note: we accept having the modules listed on the next (third) line if there is too much, using a star in the tittle. There is no well defined guideline for that AFAIK but here is an [exemple](https://github.com/odoo/enterprise/commit/7704fe0ec99950cd7404fbd7b32965bfc64ed0c4):
> 
> ```
> [FIX] hr_*: fix post-always edit
>     
>     *: hr_appraisal, hr_referral, l10n_be_hr_payroll
> ```
> 
> Or like [this](https://github.com/odoo/enterprise/commit/7785843ed0b4152b082b1254fe9c82b1fa40fa20) with a fnmatch logic
> 
> Also, if you strictly respect the line break after 72, github will detect that as a fake line break and display it in on line when generating the pr message with is convenient when you are used to it ;).

Thanks for the feedback, I modified the commit message ;)

**Comment by Xavier-Do** [comment]
@jco-odoo can you merge this one if it is ok for you?

File: addons/l10n_fr_fec/tests/test_wizard.py
Comment (william-andre) [inline addons/l10n_fr_fec/tests/test_wizard.py:25]: Why not simply decorating the method?

**Review by william-andre — COMMENTED** [review]
> This is because `TestAccountFrFec` is decorated with `freeze_time('2021-05-02')`. Indeed, when one decorates a subclass of `unittest.TestCase` with freeze_time, if the `setUpClass` raises an Exception, `_freeze_time.stop()` is not called, and the frozen date is propagated in the other test classes. As a consequence, when an exception was raised in the `setUpClass` of `TestAccountFrFec`, the freeze_time set in `TestAccountFrFec` (2021-05-02) was kept in `TestItEdi`, thus causing tests to fail in `TestItEdi`.
>
> As a conclusion, `freeze_time` should never decorate an Odoo test class.

I don't agree with the conclusion. There is a bug in `freezegun` that should be fixed.
I can do the PR on their repo if you want.
In the meantime, decoration the setupclass is fine but I don't really think that it is needed if we fix the bug that was raising an exception in the first place.

**Comment by JulienVR** [comment]
> > This is because `TestAccountFrFec` is decorated with `freeze_time('2021-05-02')`. Indeed, when one decorates a subclass of `unittest.TestCase` with freeze_time, if the `setUpClass` raises an Exception, `_freeze_time.stop()` is not called, and the frozen date is propagated in the other test classes. As a consequence, when an exception was raised in the `setUpClass` of `TestAccountFrFec`, the freeze_time set in `TestAccountFrFec` (2021-05-02) was kept in `TestItEdi`, thus causing tests to fail in `TestItEdi`.
> > As a conclusion, `freeze_time` should never decorate an Odoo test class.
> 
> I don't agree with the conclusion. There is a bug in `freezegun` that should be fixed. I can do the PR on their repo if you want. In the meantime, decoration the setupclass is fine but I don't really think that it is needed if we fix the bug that was raising an exception in the first place.

Hum... as you wish. Could then r+ the PR of JCO which is the same as this one apart from the freeze_time stuff (https://github.com/odoo/odoo/pull/108678) ?

**Comment by Xavier-Do** [comment]
> I don't agree with the conclusion. There is a bug in `freezegun` that should be fixed. I can do the PR on their repo if you want. In the meantime, decoration the setupclass is fine but I don't really think that it is needed if we fix the bug that was raising an exception in the first place.

- A fix won't appear in the current freezegun version and it will maybe take years before we get the version with the fix
- Freezetime is trying to stay cross python version compatible and  a good fix would would need to use addClassCleanup but this is not available before python 3.8 (AFAIR). We made and ovveride in odoo to have the feature in other versions.
- Other fixes are possible without that, I thought that something like that could work.

```python
  @classmethod
            def setUpClass(cls):
                self.start()
                raised = True
                try:
                    if orig_setUpClass is not None:
                        orig_setUpClass()
                    raised = False
                finally:
                    if raised:
                        self.stop()

            # noinspection PyDecorator
            @classmethod
            def tearDownClass(cls):
                try:
                    if orig_setUpClass is not None:
                        orig_setUpClass()
                finally:
                    self.stop()
 ```
 - adding it on the class adds more bugs that were fixed in next freezetime version (we are not using yet). Just to say it is a bad idea from the beginning.
- adding the freeze time on the method or with a decorator will work just fine. It is also possible to start freezetime and add a class cleanup manually to have it working on the class.
- We are thinking about replacing  freezegun freezetime calls since they causes issues for the test perf counters, profiler, logger, missing cursor time patcher, .... to replace it with a custom freezetime with correct module blacklist and additional  needs specific to odoo. And I'm not sure we will manage class decorator in this case anyway.

You can fix it with a pr in the freezegun repo if you want.  But this is not a fix for odoo ;) The fix is to remove freezetime class decorators. The conclusion is correct.

**Comment by Xavier-Do** [comment]
@robodoo r+

**Comment by william-andre** [comment]
> A fix won't appear in the current freezegun version and it will maybe take years before we get the version with the fix

No, freezegun is not part of the dependencies. We can get it as soon as it is released.

> You can fix it with a pr in the freezegun repo if you want. But this is not a fix for odoo ;) The fix is to remove freezetime class decorators. The conclusion is correct.

We don't have a bug in Odoo. It is only making some noise when there is a bug somewhere else. It doesn't need fixing.

**Comment by Xavier-Do** [comment]
> It is only making some noise when there is a bug somewhere else. It doesn't need fixing.

This is maybe not a problem for you but this is one for automated error management. We spent a few hours assigning the error trying to reproduce it and the understanding the issue because of that.

> We can get it as soon as it is released.

Since we are trying to make odoo testable without any requirement, requirements are based on debian package versions. This means that the fix won't be available before ubuntu 24 in two years.

**Pull request closed without merging**
